[#3606] Allow plugins to react to forks of the CKAN process

torfsen opened the pull request:
This is a WIP to fix #3606.

As discussed in #3606, the PR introduces a new interface (`IForkObserver`) to allow extensions to react to forks of the CKAN process. This is then used to prevent a sharing of DataStore's SQLAlchemy engines between parent and child processes, similar to what we're already doing for the core's engines.

In its current state, this PR fixes the problem of #3606 for me: if I manually enable the `datastore` plugin in `test-core.ini` the tests still succeed (without the PR some tests in `ckan/tests/lib/test_jobs.py` fail).

However, some parts of the PR are less than elegant, so I'd love to get some feedback and suggestions for improvements:

- In `ckan/lib/jobs.py` I monkey-patch `os.fork` to make sure that the implementations of `IForkObserver` are correctly notified. The monkey-patch is only temporarily while calling the `rq` method that performs the fork. An alternative would be to re-implement the `rq` method in question with the `IForkObserver` notifications. That would avoid the monkey-patch but would make updating `rq` more difficult since we would be messing even more with its internals than we already are.

- AFAIK only plugins can implement interfaces. This means that the `ckanext.datastore.plugin.Datastore` implements `IForkObserver` and then passes these events on to the active backend. Hence `ckanext.datastore.backend.DatastoreBackend` gets two new methods `before_fork` and `after_fork`, and `ckanext.datastore.backend.postgres.DatastorePostgresqlBackend` gets a new `before_fork` method. This is a lot of redirection and certainly not ideal. An slightly better alternative might be to keep these methods out of `DatastoreBackend` and then do some type checking in `Datastore.before_fork` to see whether the active backend is PostgreSQL (and if so, call its `before_fork` method).

In addition to these issues the PR is missing proper documentation. In addition, we haven't so far succeeded in creating a reproducibly failing test case for the original issue that doesn't rely on adding `datastore` to the list of plugins in `test-core.ini` (somehow, temporarily loading the plugin during the test case does not trigger the problem).

amercader commented:
@torfsen thanks for your quick work on this.

Regarding the two main points:

* Why do we need to go down to the level of `os.fork()`? To solve the main engine disposal issues on #3247 you added the environment refresh code in the [`main_work_horse`](https://github.com/torfsen/ckan/blob/df570b9d74bbe9ffd44bbb4846a19929b0aad6a7/ckan/lib/jobs.py#L254) method, which as you mention on the comment gets called shortly after forking. Can we not add the plugins check at this point so it has the same effect on extensions? eg:

    ```python

    def main_work_horse(self, job, queue):
        # This method is called in a worker's work horse process right
        # after forking.
        load_environment(config[u'global_conf'], config)
        for plugin in plugins.PluginImplementations(plugins.IForkObserver):
           plugin.after_fork(self.pid)
        return super(Worker, self).main_work_horse(job, queue)
    ```
   It could well be that I'm missing something but this seems cleaner, even if we don't have a `before_fork` method (maybe we can implement it similarly from another of the `Worker` methods).

* I agree is not ideal. Perhaps something like your last suggestion will be simpler:

    ```python

     class DatastorePlugin(p.SingletonPlugin):

          p.implements(IForkObserver)

         def after_fork(self, pid):
             if hasattr(self.backend, 'after_fork'):
                 self.backend.after_fork(pid)

     # ...

     class DatastorePostgresqlBackend(DatastoreBackend):

          def after_fork(self, pid):
                # Whatever
    ```

   On the other hand, having it explicitly added to `DatastoreBackend` kind of documents the methods (it should say these are optional). I don't have a strong opinion either way.

torfsen commented:
Consensus in the dev meeting today:
- The `after_fork` is currently not used but adds a lot of complexity. Remove it. We can add it later on if it turns out that it would be useful.
- Instead of polluting the `DatastoreBackend` interface with the events from `IForkObserver` use the duck-typing approach suggested by @amercader. A possible future improvement would be to make Datastore backends true plugins which can implement interfaces themselves.

torfsen commented:
I've implemented the changes discussed above (remove `after_fork` and propagate `before_fork` via duck-typing).

In addition I've added a test case which reproduces the problem: the test [fails](https://circleci.com/gh/ckan/ckan/2777) without the fix (dragging a lot of other test cases down with it by messing up the database connection) and [succeeds](https://circleci.com/gh/ckan/ckan/2779) when the fix is in place.

Note that GitHub displays the commits in their chronological order, so it appears that the last commit of the PR fails. However, while f46c80f is technically the youngest commit in this branch it is logically the first one. This is because I re-ordered the commits (`git rebase -i`) to put the test case to the front so that I could verify that it fails on CircleCI before pushing the fix. I wasn't aware that this would turn out to be this confusing in the GitHub UI:

```
$ git log --pretty=oneline -n 4
4d39e1e052e8047aa530f3e52596b9cbc2c748ca [#3606] Use duck-typing to call before_fork in the PostgreSQL backend
05a28a598b925bc0c6bf60736e9aae71387bb943 [#3606] Remove IForkObserver.after_fork
2e40ceb436a5164653777d1801c9896636c3129b [#3606] Allow plugins to react to forks of the CKAN process
f46c80fa34e29f3873c7110f9d594c37e0e31f6c [#3606] Add test case for Datastore backend jobs problem
```

This PR is now ready for review.

wardi commented:
@torfsen that's pretty funny.  too bad github doesn't handle rewritten history well :smile:   Having that test to demonstrate the issue is great.

amercader commented:
Thanks @torfsen, this is now on 2.7
